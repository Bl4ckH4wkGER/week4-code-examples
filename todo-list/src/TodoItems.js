import React from 'react';
import PropTypes from 'prop-types';

class TodoItems extends React.Component {
    static propTypes = {
        todos: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired,
                isComplete: PropTypes.bool.isRequired
            })
        ).isRequired,
        markComplete: PropTypes.func.isRequired
    }

    render() {
        const { todos, markComplete } = this.props;

        return (
            <ul>
                {todos.map((todo, idx) => (
                    <li
                        // className={todo.isComplete ? 'todo-done' : ''}
                        // className={todo.isComplete && 'todo-done'}
                        style={{
                            textDecoration: todo.isComplete ? 'line-through': 'inherit',
                            cursor: 'pointer'
                        }}
                        onClick={markComplete(idx)}
                        key={idx}
                    >{todo.name}</li>
                ))}
            </ul>
        )
    }
}

export default TodoItems;