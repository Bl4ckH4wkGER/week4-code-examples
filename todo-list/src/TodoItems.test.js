import React from 'react';
import { shallow, mount, render } from 'enzyme';
import TodoItems from './TodoItems';

test('addition', () => {
    expect(2 + 2).toEqual(4);
});

test('TodoItems renders with no todos', () => {
    const markComplete = jest.fn();
    const element = render(<TodoItems todos={[]} markComplete={markComplete} />);

    expect(element).toMatchSnapshot();
});

test('TodoItems renders with a todo', () => {
    const markComplete = jest.fn();
    const todos = [{ name: 'Walk the dog', isComplete: false }];
    const element = render(<TodoItems todos={todos} markComplete={markComplete} />);

    expect(element).toMatchSnapshot();
});

test('TodoItems correctly marks a completed item with a line-through', () => {
    const markComplete = jest.fn();
    const todos = [{ name: 'Walk the dog', isComplete: true }];
    const element = shallow(<TodoItems todos={todos} markComplete={markComplete} />);

    const listItems = element.find('li');
    const style = listItems.get(0).props.style;

    expect(element).toMatchSnapshot();
    expect(style.textDecoration).toEqual('line-through');
});

test('Clicking on a todo item calls markComplete', () => {
    const markCompleteEventHandler = jest.fn();
    const markComplete = jest.fn(() => markCompleteEventHandler);

    const todos = [{ name: 'Walk the dog', isComplete: true }];
    const element = shallow(<TodoItems todos={todos} markComplete={markComplete} />);

    expect(markComplete).toHaveBeenCalledTimes(1);
    expect(markComplete).toHaveBeenCalledWith(0); // index 0
    expect(markCompleteEventHandler).not.toHaveBeenCalled();

    const listItem = element.find('li').first();
    listItem.simulate('click');

    expect(markCompleteEventHandler).toHaveBeenCalledTimes(1);
});