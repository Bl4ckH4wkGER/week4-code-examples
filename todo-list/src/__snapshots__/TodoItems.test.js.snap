// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItems correctly marks a completed item with a line-through 1`] = `
<ul>
  <li
    key="0"
    style={
      Object {
        "cursor": "pointer",
        "textDecoration": "line-through",
      }
    }
  >
    Walk the dog
  </li>
</ul>
`;

exports[`TodoItems renders with a todo 1`] = `
<ul>
  <li
    style="text-decoration:inherit;cursor:pointer"
  >
    Walk the dog
  </li>
</ul>
`;

exports[`TodoItems renders with no todos 1`] = `<ul />`;
